### PR TITLE
Differentiable local field projection

### LINKED ISSUE #1114: adjoint Field projection monitor support

### PR DESCRIPTION
**Pull request by yaugenst-flex**

### CODE REVIEW

**Comment by tylerflex** [comment]
👀

File: CHANGELOG.md
Comment (tylerflex) [inline CHANGELOG.md]: could you provide more detail on the specific call(s)? for example, include the name of the method(s) that are differentiable now? `FieldProjector(fld_mnt, ...).project(...)`

File: tidy3d/components/autograd/utils.py
Comment (tylerflex) [inline tidy3d/components/autograd/utils.py]: docstrings here (and other functions)

File: tidy3d/components/base.py
Comment (tylerflex) [inline tidy3d/components/base.py]: hm.. so just curious. if x is a DataArray, will this already just trigger this condition and then assign the `field_mapping[path]`? I'm worried that this logic will never reach `elif isinstance(x, DataArray)` when x is traced.

**Review by tylerflex — COMMENTED** [review]
a few comments. I'm mostly a little confused about the base.py stuff?

File: tidy3d/components/autograd/utils.py
Comment (tylerflex) [inline tidy3d/components/autograd/utils.py]: `is_box` and `is_traced` look better to me here. Basically we're just adding a special case to handle `np.ndarray` with dtype=object and `ArrayBox` elements?

File: tidy3d/components/base.py
Comment (tylerflex) [inline tidy3d/components/base.py]: remove comment, handled in `insert_tracers`

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: remove comment

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: note: this inserts the dtype=object into the `values`, as we discussed. we should revisit this choice later.

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: note: we could consider just adding this to a `self.autograd_tracers` attribute instead of attrs. not sure why I used attrs originally.

File: tidy3d/components/autograd/utils.py
Comment (tylerflex) [inline tidy3d/components/autograd/utils.py]: should we consider just calling this `get_tracer`? to be consistent with our naming?

Comment (tylerflex) [reply tidy3d/components/autograd/utils.py]: or not sure if there's a better word for these things, maybe just use `autograd_box` or `box` everywhere?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: put docstring back?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: docstrings and type annotations for these properties.

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: does `getval` need a wrapper in `utils.py`? I'm not 100% sure about what happens here if `self.values` is a `np.ndarray` of `dtype=object`. my impression is that getval will not work here unless we apply it to `np.ndarray(self.values.tolist())` but could be wrong.

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: might be better to move right under init to keep it more organized

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: I think we can use `insert_tracers` here instead. And then modify that method to handle the `attrs.pop(AUTOGRAD_KEY, None)`

We might also want to change the name of that method to reflect that we are not necessarily inserting tracers only. Maybe a more accurate name for it is `.update_with_data()`

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: we do have a `to_static()` in the `Tidy3dBaseModel`. Maybe one or both of those should change to be consistent with each other?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: I feel like this could be its own method, or maybe it should use the `getbox` util?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: I think here we can use `self.insert_tracers(vals)`?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: probably we can refactor this. or at least use `get_box` from utils? and for real as well

**Review by tylerflex — COMMENTED** [review]
Some comments.. Mostly things I thought could be refactored, renamed, simplified. 

Otherwise I think it's consistent / correct. Only thing I'm not 100% sure on is in the `DataArray.static` property.

```py
vals = self.values if self.values.size == 0 else np.vectorize(getval)(self.values)

```
will `getval` work if `self.values` is a np.ndarray of dtype=object?

otherwise feeling pretty good about it in its current state

File: tidy3d/components/autograd/utils.py
Comment (yaugenst-flex) [reply tidy3d/components/autograd/utils.py]: Yeah exactly, this got refactored to the point where it probably doesn't make sense to have anymore :laughing:
If we don't do it here, we'll still need to handle object arrays somewhere else though.

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [reply tidy3d/components/data/data_array.py]: ah, maybe the vectorize already takes care of this.

File: tidy3d/components/autograd/utils.py
Comment (yaugenst-flex) [reply tidy3d/components/autograd/utils.py]: I just used box because that's the autograd terminology. Either is fine. Tracer is probably less confusing though if we stick with one name.

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py]: Oops. Although, to be fair, this is like
```python
# initialize x
x = 0
```

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py]: yeah the vectorize is there exactly for that purpose. `np.array(self.values.tolist())` doesn't work actually because `getval` will not strip tracers from nested lists, i.e. if the array is >1d then the `tolist` approach will not work

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py]: let's pick `to_static` then

File: tidy3d/components/autograd/utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/utils.py]: yea it makes sense, but maybe we can just use one word for this
* `tracer` = `ArrayBox / Box`
* `traced` = contains a `tracer`
* `get_tracer` = grab the `tracer` from something
* `untraced` = copy of something without the tracer
* `insert_tracer` = copy of something with the tracer inserted.
?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [reply tidy3d/components/data/data_array.py]: ah, good to know. Does this pop up anywhere else? (ie where the `anp.array(...tolist())` was called previously?) if so, maybe we should refactor it into one helper function

Comment (tylerflex) [reply tidy3d/components/data/data_array.py]: sorry, got a bit confused.. never mind

File: tidy3d/components/autograd/utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/utils.py]: object arrays only pop up in the DataArray I think, so maybe we can just handle it there? unless it makes it easier to do it here.

**Comment by yaugenst-flex** [comment]
@tylerflex There are some changes to `DataArray` properties here. The naming and mechanics of these are probably not ideal, but since it's not user facing let's address them in the `DataArray` refactor? We'll end up changing most of these things anyways.
Also, #1939 lists this for `2.8rc1`, should I rebase on `pre/2.8`? Although I think it makes sense to include this only with the gradient fixes in `develop`. So either as part of `2.7.3` or rebase to `pre/2.8` only after `develop` is merged.

**Comment by tylerflex** [comment]
> Also, https://github.com/flexcompute/tidy3d/issues/1939 lists this for 2.8rc1, should I rebase on pre/2.8? Although I think it makes sense to include this only with the gradient fixes in develop. So either as part of 2.7.3 or rebase to pre/2.8 only after develop is merged.

@momchil-flex what do you think? basically there were some changes in my (recently merged) #1923 that Yannick used in this PR. So ideally we would do another develop -> pre/2.8 merge before Yannick rebases against 2.8?

**Comment by momchil-flex** [comment]
@yaugenst-flex what do you think will be simplest? I'm already running notebook tests for 2.7.3 and since it's hard to tell if this PR may not have various small effects, I'd rather not include it for now. However we could merge it to develop after 2.7.3 and then eventually into pre/2.8, or we can merge develop into pre/2.8 again and then you rebase the PR. Either way it will likely only come out in 2.8.0rc1 since we may not do any other 2.7. patches, but if we did the first approach would put it in there too.

**Comment by yaugenst-flex** [comment]
@momchil-flex I think it makes sense either way to merge develop into pre/2.8 after the 2.7.3 release so we keep them somewhat in sync. I'll rebase to pre/2.8 after that, shouldn't be a problem.

File: pyproject.toml
Comment (tylerflex) [inline pyproject.toml]: we need to be careful about this due to the danger of bloating our core dependencies. It became an issue recently and we agreed to let MC know if we add any dependencies. Is this absolutely necessary you think?

File: tests/ruff.toml
Comment (tylerflex) [inline tests/ruff.toml:13]: snuck in from another PR?

File: tidy3d/components/autograd/utils.py
Comment (tylerflex) [inline tidy3d/components/autograd/utils.py]: pretty minor but slighly cleaner might be
```py
if _is_traced_objary(x):
     return anp.array(x.tolist())

return x
```
?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: why skip exactly? can't we interpolate tracer. containing DataArrays now?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py]: perhaps we can refactor this logic out (eg. the one in real, conj, etc).

```py
def apply_traced(self, fn):

    if isbox(self.values.flat[0]) or AUTOGRAD_KEY in self.attrs:
        tracers = anp.array(self.values.tolist()))
        new_values = fn(tracers)
        return self.insert_tracers(new_values)

    return fn(self)
    
@property
def real(self):
    return self.apply_traced(anp.real)
```

Comment (tylerflex) [reply tidy3d/components/data/data_array.py]: also it seems the `anp.array(self.values.tolist())` and `if` statement both could use the convenience methods / functions used to check these things?

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [inline tidy3d/components/data/data_array.py:1]: I am blanking on a lot of the discussions we had regarding the data array handing, but will mostly assume that this logic follows what we already talked about before?

Is there anything you think I should take a more detailed look at?

File: tidy3d/components/data/monitor_data.py
Comment (tylerflex) [inline tidy3d/components/data/monitor_data.py]: do we need to do this treatment in other parts of this class? I think this kind of operation does pop up from time to time?

File: tidy3d/plugins/autograd/README.md
Comment (tylerflex) [inline tidy3d/plugins/autograd/README.md]: was this mentioned as a "to do" item? if so would be good to remove it.

**Review by tylerflex — COMMENTED** [review]
Looks really good! thanks @yaugenst-flex . I dont know if you want @QimingFlex to look at the field projection bits? I couldn't understand most of the physics of the operations, but did follow through the code and thought it looked good.

File: pyproject.toml
Comment (yaugenst-flex) [reply pyproject.toml]: `opt-einsum` is a pretty tiny dependency (~200kb) and itself only depends on numpy, and it's already a dependency of jax (although that's not core). Otherwise I can also rewrite that part using regular numpy einsum, not a big deal.

File: tests/ruff.toml
Comment (yaugenst-flex) [reply tests/ruff.toml:13]: Actually no, this got flagged (`np.trapz` usage) and I disabled it. We should enable it again once we support numpy 2.

File: tidy3d/components/autograd/utils.py
Comment (yaugenst-flex) [reply tidy3d/components/autograd/utils.py]: good point

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py]: Yeah this probably should have been removed in the interp PR...

File: tidy3d/components/data/monitor_data.py
Comment (yaugenst-flex) [reply tidy3d/components/data/monitor_data.py]: If we want to differentiate w.r.t. those methods then yeah this needs to be done. Although I'm hoping that we can get rid of this "hack" with a DataArray refactor...

File: tidy3d/plugins/autograd/README.md
Comment (yaugenst-flex) [reply tidy3d/plugins/autograd/README.md]: not seeing that anywhere in this PR, but I know I've seen it somewhere...

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py]: Not really sure how to handle the `super()` call this way? But yeah I can at least partially refactor this..

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py:1]: Yeah I think it's fine for the scope of this PR, so no need to dive deep for now, I vote that we put together a proper DataArray refactor soonish though.

**Comment by yaugenst-flex** [comment]
Hey @QimingFlex, would be great if you could have a quick look at the field projection bits in this PR, I reworked some of it to more easily accomodate autograd. I did try to make sure that all of the field projection values that come out of this are still exactly the same as before, but it wouldn't hurt to have another pair of eyes on it.

File: tidy3d/components/data/data_array.py
Comment (tylerflex) [reply tidy3d/components/data/data_array.py]: do we need super or can `np.real(self)` suffice?

File: pyproject.toml
Comment (tylerflex) [reply pyproject.toml]: @lei-flex what do you think about adding this dependency?

File: tidy3d/components/data/data_array.py
Comment (yaugenst-flex) [reply tidy3d/components/data/data_array.py]: wouldn't that be infinite recursion? pretty sure `np.real(self)` will call `self.real`

File: tidy3d/components/field_projection.py
Comment (QimingFlex) [inline tidy3d/components/field_projection.py]: How does autograd interact with the current computation here?

**Comment by QimingFlex** [comment]
Hi @yaugenst-flex ,looks good to me, just a few questions regarding the surface current computation.

File: tidy3d/components/field_projection.py
Comment (QimingFlex) [inline tidy3d/components/field_projection.py]: I'm finding this part challenging to understand. The original code followed Taflove's textbook, which was more comprehensible. Would it be beneficial to add some explanatory comments here to clarify the logic and purpose of these operations?

File: tidy3d/components/field_projection.py
Comment (yaugenst-flex) [reply tidy3d/components/field_projection.py]: this is related to how the traced values are stored. essentially, they can either be in `AUTOGRAD_KEY` in the `.attrs` dictionary, or they might be stored in the `DataArray`'s `.values` directly. So this line just extracts those. When no autograd is involved, this is equivalent to just calling `currents[eh_key].values`.

File: tidy3d/components/field_projection.py
Comment (yaugenst-flex) [reply tidy3d/components/field_projection.py]: hey @QimingFlex sorry for the late reply on this. yeah I understand this is quite a big difference to how it was implemented previously. the calculations are the same, so it's still following taflove, but the implementation is obviously quite different.
this is because the calculation is now vectorized over theta and phi. the reason for this is mainly that this is much more efficient during the backward pass, otherwise the memory requirements can go through the roof when looping over each array element.
without gradients, this implementation is around as fast as the original implementation for a small number of theta, and faster for a large number of theta.

because of this vectorization, the einsum makes it much simpler to keep track of the dimensions in those arrays (i.e., it is clearly labeled which dimensions remain, which ones disappear / get summed over). but I understand your concern that this is less readable in terms of just following the computation.

do you think we should make this compromise or should I try to refactor it a bit?

Comment (QimingFlex) [reply tidy3d/components/field_projection.py]: Thanks, @yaugenst-flex. I've reviewed the operations, and they look good to me! It's great to see some acceleration achieved for the frontend Near2far. I do have one suggestion: it might be better to keep with the original abbreviations, such as `phase` instead of `ehp`, as the former is more intuitive. The same applies to terms like `st` and `ct`. Although the original names are longer, they are generally easier to follow.

File: pyproject.toml
Comment (yaugenst-flex) [reply pyproject.toml]: I removed the opt-einsum dependency, using numpy instead

File: tidy3d/components/field_projection.py
Comment (yaugenst-flex) [reply tidy3d/components/field_projection.py]: thanks! yeah that's a good point, I implemented your suggestions as well as some other readability improvements as well

**Review by QimingFlex — APPROVED** [review]

**Review by tylerflex — APPROVED** [review]

**Pull request merged**
